src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java

```diff
@@ -198,20 +362,20 @@ Result getResultForParticipation(Participation participation, ZonedDateTime dueD
                 return ratedResults.get(0);
             }
 
-            // sorting in descending order to have the last result at the beginning
-            resultsList.sort(Comparator.comparing(Result::getCompletionDate).reversed());
+            Result latestResult = Collections.max(resultsList, Comparator.comparing(Result::getCompletionDate));
 
             long gracePeriodInSeconds = 10L;
-            if (dueDate == null || !dueDate.plusSeconds(gracePeriodInSeconds).isBefore(resultsList.get(0).getCompletionDate())) {
+            if (dueDate == null || (latestResult.getCompletionDate() != null && !dueDate.plusSeconds(gracePeriodInSeconds).isBefore(latestResult.getCompletionDate()))) {
                 // find the first result that is before the due date
-                chosenResult = resultsList.get(0);
+                chosenResult = latestResult;
             }
-            else if (dueDate.plusSeconds(gracePeriodInSeconds).isBefore(resultsList.get(0).getCompletionDate())) {
+            else if (latestResult.getCompletionDate() == null || dueDate.plusSeconds(gracePeriodInSeconds).isBefore(latestResult.getCompletionDate())) {
                 chosenResult = new Result();
                 chosenResult.setScore(0.0);
             }
             else {
-                chosenResult = resultsList.get(resultsList.size() - 1);
+                // Choose the oldest result.
+                chosenResult = Collections.min(resultsList, Comparator.comparing(Result::getCompletionDate));
```

Comment:
Also not sure if this is correct, I need to think about this again.

Reply:
I changed the method to look for the latest rated result before the due date, and if there is none, a 0 Result is returned. I really don't get why you would just return the oldest result.